dbms/src/Operators/AutoPassThroughHashAggContext.cpp

```diff
@@ -0,0 +1,198 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Columns/ColumnDecimal.h>
+#include <Operators/AutoPassThroughHashAggContext.h>
+
+#include <ext/scope_guard.h>
+#include <magic_enum.hpp>
+
+namespace DB
+{
+void AutoPassThroughHashAggContext::onBlockAuto(Block & block)
+{
+    RUNTIME_CHECK_MSG(!already_start_to_get_data, "Shouldn't insert into HashMap if already start to get data");
+
+    agg_process_info->resetBlock(block);
+    switch (state)
+    {
+    case State::Init:
+    {
+        aggregator->executeOnBlock(*agg_process_info, *many_data[0], 0);
+        trySwitchFromInitState();
+        break;
+    }
+    case State::Adjust:
+    {
+        aggregator->executeOnBlockCollectHitRate(*agg_process_info, *many_data[0], 0);
+        trySwitchFromAdjustState(agg_process_info->block.rows(), agg_process_info->hit_row_cnt);
+        break;
+    }
+    case State::PreHashAgg:
+    {
+        aggregator->executeOnBlock(*agg_process_info, *many_data[0], 0);
+        trySwitchBackAdjustState(agg_process_info->block.rows());
+        break;
+    }
+    case State::PassThrough:
+    {
+        const auto total_rows = agg_process_info->block.rows();
+        onBlockForceStreaming(agg_process_info->block);
+        trySwitchBackAdjustState(total_rows);
+        break;
+    }
+    case State::Selective:
+    {
+        aggregator->executeOnBlockOnlyLookup(*agg_process_info, *many_data[0], 0);
+        auto pass_through_rows = agg_process_info->not_found_rows;
+        const auto total_rows = agg_process_info->block.rows();
+        if (!pass_through_rows.empty())
+        {
+            RUNTIME_CHECK(!agg_process_info->block.info.selective);
+            auto new_block = getPassThroughBlock(agg_process_info->block);
+            new_block.info.selective = std::make_shared<std::vector<UInt64>>(std::move(pass_through_rows));
+            pushPassThroughBuffer(new_block);
+        }
+        trySwitchBackAdjustState(total_rows);
+        break;
+    }
+    default:
+    {
+        __builtin_unreachable();
+    }
+    };
+    RUNTIME_CHECK(agg_process_info->allBlockDataHandled());
+}
+
+void AutoPassThroughHashAggContext::onBlockForceStreaming(Block & block)
+{
+    auto new_block = getPassThroughBlock(block);
+    pushPassThroughBuffer(new_block);
+}
+
+void AutoPassThroughHashAggContext::forceState()
+{
+    if (many_data[0]->need_spill)
+        state = State::PassThrough;
```

Comment:
will the hash table be converted to block after check `need_spill` is true?

Reply:
When `need_spill` is true, follow will happend:
1. stop insert new data into hashtable
2. will convert hashtable to block in advance(`getDataInAdvance()`) to help reduce memory.